File: libpyclingo/clingo/solving.py
"""
Functions and classes related to solving.

Examples
--------

The following example shows how to intercept models with a callback:

    >>> from clingo import Control
    >>>
    >>> ctl = Control(["0"])
    >>> ctl.add("base", [], "1 { a; b } 1.")
    >>> ctl.ground([("base", [])])
    >>> print(ctl.solve(on_model=print))
    Answer: a
    Answer: b
    SAT

The following example shows how to yield models:

    >>> from clingo import Control
    >>>
    >>> ctl = Control(["0"])
    >>> ctl.add("base", [], "1 { a; b } 1.")
    >>> ctl.ground([("base", [])])
    >>> with ctl.solve(yield_=True) as hnd:
    ...     for m in hnd:
    ...         print(m)
    ...     print(hnd.get())
    ...
    Answer: a
    Answer: b
    SAT

The following example shows how to solve asynchronously:

    >>> from clingo import Control
    >>>
    >>> ctl = Control(["0"])
    >>> ctl.add("base", [], "1 { a; b } 1.")
    >>> ctl.ground([("base", [])])
    >>> with ctl.solve(on_model=print, async_=True) as hnd:
    ...     # some computation here
    ...     hnd.wait()
    ...     print(hnd.get())
    ...
    Answer: a
    Answer: b
    SAT

This example shows how to solve both iteratively and asynchronously:

    >>> from clingo import Control
    >>>
    >>> ctl = Control(["0"])
    >>> ctl.add("base", [], "1 { a; b } 1.")
    >>> ctl.ground([("base", [])])
    >>> with ctl.solve(yield_=True, async_=True) as hnd:
    ...     while True:
    ...         hnd.resume()
    ...         # some computation here
    ...         _ = hnd.wait()
    ...         m = hnd.model()
    ...         if m is None:
    ...             print(hnd.get())
    ...             break
    ...         print(m)
    b
    a
    a b
    None
"""

from typing import ContextManager, Iterator, List, Optional, Sequence, Tuple, Union

from ._internal import _c_call, _c_call2, _ffi, _handle_error, _lib
from .util import Slice, SlicedSequence
from .core import OrderedEnum
from .symbol import Symbol
from .symbolic_atoms import SymbolicAtoms

__all__ = ["Model", "ModelType", "SolveControl", "SolveHandle", "SolveResult"]


class SolveResult:
    """
    Captures the result of a solve call.
    """

    def __init__(self, rep):
        self._rep = rep

    @property
    def exhausted(self) -> bool:
        """
        Determine if the search space was exhausted.
        """
        return (_lib.clingo_solve_result_exhausted & self._rep) != 0

    @property
    def interrupted(self) -> bool:
        """
        Determine if the search was interrupted.
        """
        return (_lib.clingo_solve_result_interrupted & self._rep) != 0

    @property
    def satisfiable(self) -> Optional[bool]:
        """
        `True` if the problem is satisfiable, `False` if the problem is
        unsatisfiable, or `None` if the satisfiablity is not known.
        """
        if (_lib.clingo_solve_result_satisfiable & self._rep) != 0:
            return True
        if (_lib.clingo_solve_result_unsatisfiable & self._rep) != 0:
            return False
        return None

    @property
    def unknown(self) -> bool:
        """
        Determine if the satisfiablity is not known.

        This is equivalent to satisfiable is None.
        """
        return self.satisfiable is None

    @property
    def unsatisfiable(self) -> Optional[bool]:
        """
        `True` if the problem is unsatisfiable, `False` if the problem is
        satisfiable, or `None` if the satisfiablity is not known.
        """
        if (_lib.clingo_solve_result_unsatisfiable & self._rep) != 0:
            return True
        if (_lib.clingo_solve_result_satisfiable & self._rep) != 0:
            return False
        return None

    def __str__(self):
        if self.satisfiable:
            return "SAT"
        if self.unsatisfiable:
            return "UNSAT"
        return "UNKNOWN"

    def __repr__(self):
        return f"SolveResult({self._rep})"


class SolveControl:
    """
    Object that allows for controlling a running search.
    """

    def __init__(self, rep):
        self._rep = rep

    def add_clause(self, literals: Sequence[Union[Tuple[Symbol, bool], int]]) -> None:
        """
        Add a clause that applies to the current solving step during the search.

        Parameters
        ----------
        literals
            List of literals either represented as pairs of symbolic atoms and
            Booleans or as program literals.

        Notes
        -----
        This function can only be called in a model callback or while iterating
        when using a `SolveHandle`.
        """
        atoms = self.symbolic_atoms
        p_lits = _ffi.new("clingo_literal_t[]", len(literals))
        for i, lit in enumerate(literals):
            if isinstance(lit, int):
                p_lits[i] = lit
            else:
                atom = atoms[lit[0]]
                if atom is not None:
                    slit = atom.literal
                else:
                    slit = -1
                p_lits[i] = slit if lit[1] else -slit

        _handle_error(
            _lib.clingo_solve_control_add_clause(self._rep, p_lits, len(literals))
        )

    def _invert(
        self, lit: Union[Tuple[Symbol, bool], int]
    ) -> Union[Tuple[Symbol, bool], int]:
        if isinstance(lit, int):
            return -lit
        return lit[0], not lit[1]

    def add_nogood(self, literals: Sequence[Union[Tuple[Symbol, bool], int]]) -> None:
        """
        Equivalent to `SolveControl.add_clause` with the literals inverted.
        """
        self.add_clause([self._invert(lit) for lit in literals])

    @property
    def symbolic_atoms(self) -> SymbolicAtoms:
        """
        `clingo.symbolic_atoms.SymbolicAtoms` object to inspect the symbolic atoms.
        """
        atoms = _c_call(
            "clingo_symbolic_atoms_t*",
            _lib.clingo_solve_control_symbolic_atoms,
            self._rep,
        )
        return SymbolicAtoms(atoms)


class ModelType(OrderedEnum):
    """
    Enumeration of the different types of models.
    """

    BraveConsequences = _lib.clingo_model_type_brave_consequences
    """
    The model stores the set of brave consequences.
    """
    CautiousConsequences = _lib.clingo_model_type_cautious_consequences
    """
    The model stores the set of cautious consequences.
    """
    StableModel = _lib.clingo_model_type_stable_model
    """
    The model captures a stable model.
    """


class _SymbolSequence(Sequence[Symbol]):
    """
    Helper class to efficiently store sequences of symbols.
    """

    def __init__(self, p_symbols):
        self._p_symbols = p_symbols

    def __len__(self):
        return len(self._p_symbols)

    def __getitem__(self, slc):
        if isinstance(slc, slice):
            return SlicedSequence(self, Slice(slc))
        if slc < 0:
            slc += len(self)
        if slc < 0 or slc >= len(self):
            raise IndexError("invalid index")
        return Symbol(self._p_symbols[slc])

    def __iter__(self):
        for i in range(len(self)):
            yield Symbol(self._p_symbols[i])

    def __str__(self):
        return f'[{", ".join(str(sym) for sym in self)}]'

    def __repr__(self):
        return f'[{", ".join(repr(sym) for sym in self)}]'


class Model:
    """
    Provides access to a model during a solve call and provides a
    `SolveContext` object to influence the running search.

    Notes
    -----
    The string representation of a model object is similar to the output of
    models by clingo using the default output.

    `Model` objects cannot be constructed from Python. Instead they are obained
    during solving (see `Control.solve`). Furthermore, the lifetime of a model
    object is limited to the scope of the callback it was passed to or until
    the search for the next model is started. They must not be stored for later
    use.
    """

    def __init__(self, rep):
        self._rep = rep

    def contains(self, atom: Symbol) -> bool:
        """
        Efficiently check if an atom is contained in the model.

        Parameters
        ----------
        atom
            The atom to lookup.

        Returns
        -------
        Whether the given atom is contained in the model.

        Notes
        -----
        The atom must be represented using a function symbol.
        """
        # pylint: disable=protected-access
        return _c_call("bool", _lib.clingo_model_contains, self._rep, atom._rep)

    def extend(self, symbols: Sequence[Symbol]) -> None:
        """
        Extend a model with the given symbols.

        Parameters
        ----------
        symbols
            The symbols to add to the model.

        Notes
        -----
        This only has an effect if there is an underlying clingo application,
        which will print the added symbols.
        """
        # pylint: disable=protected-access
        c_symbols = _ffi.new("clingo_symbol_t[]", len(symbols))
        for i, sym in enumerate(symbols):
            c_symbols[i] = sym._rep
        _handle_error(_lib.clingo_model_extend(self._rep, c_symbols, len(symbols)))

    def is_true(self, literal: int) -> bool:
        """
        Check if the given program literal is true.

        Parameters
        ----------
        literal
            The given program literal.

        Returns
        -------
        Whether the given program literal is true.
        """
        return _c_call("bool", _lib.clingo_model_is_true, self._rep, literal)

    def is_consequence(self, literal: int) -> Optional[bool]:
        """
        Check if the given program literal is a consequence.

        The function returns `True`, `False`, or `None` if the literal is a
        consequence, not a consequence, or it is not yet known whether it is a
        consequence, respectively.

        While enumerating cautious or brave consequences, there is partial
        information about which literals are consequences. The current state of
        a literal can be requested using this function. If this function is
        used during normal model enumeration, the function just returns whether
        a literal is true of false in the current model.

        Parameters
        ----------
        literal
            The given program literal.

        Returns
        -------
        Whether the given program literal is a consequence.
        """
        res = _c_call(
            "clingo_consequence_t", _lib.clingo_model_is_consequence, self._rep, literal
        )
        if res == _lib.clingo_consequence_true:
            return True
        if res == _lib.clingo_consequence_false:
            return False
        return None

    def symbols(
        self,
        atoms: bool = False,
        terms: bool = False,
        shown: bool = False,
        theory: bool = False,
        complement: bool = False,
    ) -> Sequence[Symbol]:
        """
        Return the list of atoms, terms, or CSP assignments in the model.

        Parameters
        ----------
        atoms
            Select all atoms in the model (independent of `#show` statements).
        terms
            Select all terms displayed with `#show` statements in the model.
        shown
            Select all atoms and terms as outputted by clingo.
        theory
            Select atoms added with `Model.extend`.
        complement
            Return the complement of the answer set w.r.t. to the atoms known
            to the grounder.

        Returns
        -------
        The selected symbols.

        Notes
        -----
        Atoms are represented using functions (`Symbol` objects), and CSP
        assignments are represented using functions with name `"$"` where the
        first argument is the name of the CSP variable and the second its
        value.
        """
        show = 0
        if atoms:
            show |= _lib.clingo_show_type_atoms
        if terms:
            show |= _lib.clingo_show_type_terms
        if shown:
            show |= _lib.clingo_show_type_shown
        if theory:
            show |= _lib.clingo_show_type_theory
        if complement:
            show |= _lib.clingo_show_type_complement

        size = _c_call("size_t", _lib.clingo_model_symbols_size, self._rep, show)

        p_symbols = _ffi.new("clingo_symbol_t[]", size)
        _handle_error(_lib.clingo_model_symbols(self._rep, show, p_symbols, size))

        return _SymbolSequence(p_symbols)

    def __str__(self):
        return " ".join(map(str, self.symbols(shown=True)))

    def __repr__(self):
        return f"Model({self._rep!r})"

    @property
    def context(self) -> SolveControl:
        """
        Object that allows for controlling the running search.
        """
        ctl = _c_call("clingo_solve_control_t*", _lib.clingo_model_context, self._rep)
        return SolveControl(ctl)

    @property
    def cost(self) -> List[int]:
        """
        Return the list of integer cost values of the model.

        The return values correspond to clasp's cost output.
        """
        size = _c_call("size_t", _lib.clingo_model_cost_size, self._rep)

        p_costs = _ffi.new("int64_t[]", size)
        _handle_error(_lib.clingo_model_cost(self._rep, p_costs, size))

        return list(p_costs)

    @property
    def priority(self) -> List[int]:
        """
        Return the priorities of the model's cost values.
        """
        size = _c_call("size_t", _lib.clingo_model_cost_size, self._rep)

        p_priorities = _ffi.new("clingo_weight_t[]", size)
        _handle_error(_lib.clingo_model_priority(self._rep, p_priorities, size))

        return list(p_priorities)

    @property
    def number(self) -> int:
        """
        The running number of the model.
        """
        return _c_call("uint64_t", _lib.clingo_model_number, self._rep)

    @property
    def optimality_proven(self) -> bool:
        """
        Whether the optimality of the model has been proven.
        """
        return _c_call("bool", _lib.clingo_model_optimality_proven, self._rep)

    @property
    def thread_id(self) -> int:
        """
        The id of the thread which found the model.
        """
        return _c_call("clingo_id_t", _lib.clingo_model_thread_id, self._rep)

    @property
    def type(self) -> ModelType:
        """
        The type of the model.
        """
        return ModelType(
            _c_call("clingo_model_type_t", _lib.clingo_model_type, self._rep)
        )


class SolveHandle(ContextManager["SolveHandle"]):
    """
    Handle for solve calls.

    They can be used to control solving, like, retrieving models or cancelling
    a search.

    See Also
    --------
    Control.solve

    Notes
    -----
    A `SolveHandle` is a context manager and must be used with Python's `with`
    statement.

    Blocking functions in this object release the GIL. They are not thread-safe
    though.
    """

    def __init__(self, rep, handler):
        self._rep = rep
        self._handler = handler

    def __iter__(self) -> Iterator[Model]:
        while True:
            self.resume()
            m = self.model()
            if m is None:
                break
            yield m

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        _handle_error(_lib.clingo_solve_handle_close(self._rep), self._handler)
        return False

    def cancel(self) -> None:
        """
        Cancel the running search.

        See Also
        --------
        clingo.control.Control.interrupt
        """
        _handle_error(_lib.clingo_solve_handle_cancel(self._rep), self._handler)

    def core(self) -> List[int]:
        """
        The subset of assumptions that made the problem unsatisfiable.
        """
        core, size = _c_call2(
            "clingo_literal_t*",
            "size_t",
            _lib.clingo_solve_handle_core,
            self._rep,
            handler=self._handler,
        )
        return [core[i] for i in range(size)]

    def get(self) -> SolveResult:
        """
        Get the result of a solve call.

        If the search is not completed yet, the function blocks until the
        result is ready.
        """
        res = _c_call(
            "clingo_solve_result_bitset_t",
            _lib.clingo_solve_handle_get,
            self._rep,
            handler=self._handler,
        )
        return SolveResult(res)

    def model(self) -> Optional[Model]:
        """
        Get the current model if there is any.
        """
        p_model = _ffi.new("clingo_model_t**")
        _handle_error(_lib.clingo_solve_handle_model(self._rep, p_model), self._handler)
        if p_model[0] == _ffi.NULL:
            return None
        return Model(p_model[0])

    def resume(self) -> None:
        """
        Discards the last model and starts searching for the next one.

        Notes
        -----
        If the search has been started asynchronously, this function starts the
        search in the background.
        """
        _handle_error(_lib.clingo_solve_handle_resume(self._rep), self._handler)

    def wait(self, timeout: Optional[float] = None) -> bool:
        """
        Wait for solve call to finish or the next result with an optional timeout.

        If a timeout is given, the behavior of the function changes depending
        on the sign of the timeout. If a postive timeout is given, the function
        blocks for the given amount time or until a result is ready. If the
        timeout is negative, the function will block until a result is ready,
        which also corresponds to the behavior of the function if no timeout is
        given. A timeout of zero can be used to poll if a result is ready.

        Parameters
        ----------
        timeout
            If a timeout is given, the function blocks for at most timeout seconds.

        Returns
        -------
        Indicates whether the solve call has finished or the next result is ready.
        """
        p_res = _ffi.new("bool*")
        _lib.clingo_solve_handle_wait(
            self._rep, -1 if timeout is None else timeout, p_res
        )
        return p_res[0]

File: lib/python-api/src/solving.cc
#include "solving.hh"
#include "core.hh"
#include "util.hh"

#include <clingo/solve.h>

#include <utility>

namespace Clingo::Python {

namespace {

auto symbol_callback(clingo_symbol_t const *symbols, size_t size, void *data) -> clingo_result_t {
    CLINGO_TRY {
        auto *res = static_cast<SymbolVec *>(data);
        // NOLINTNEXTLINE
        res->assign(cpp_cast(symbols), cpp_cast(symbols) + size);
    }
    CLINGO_CATCH(nullptr);
}

class ModelIterator {
  public:
    using iterator_category = std::forward_iterator_tag;
    using difference_type = std::ptrdiff_t;
    using value_type = Model;
    using pointer = Model *;
    using reference = Model &;

    ModelIterator() = default;
    ModelIterator(SSolveHandle hnd) : hnd_(std::move(hnd)) { operator++(); }

    // NOLINTBEGIN(bugprone-unchecked-optional-access)
    auto operator*() -> reference { return mdl_.value(); }
    auto operator->() -> pointer { return &mdl_.value(); }
    // NOLINTEND(bugprone-unchecked-optional-access)

    auto operator++() -> ModelIterator & {
        if (hnd_) {
            hnd_->resume();
            mdl_ = hnd_->model();
            if (!mdl_) {
                hnd_.reset();
            }
        }
        return *this;
    }
    auto operator++(int) -> ModelIterator {
        ModelIterator tmp = *this;
        ++(*this);
        return tmp;
    }

    [[maybe_unused]] friend auto operator==(ModelIterator const &a, ModelIterator const &b) -> bool {
        return a.hnd_ == b.hnd_;
    }
    [[maybe_unused]] friend auto operator!=(ModelIterator const &a, ModelIterator const &b) -> bool {
        return a.hnd_ != b.hnd_;
    }

  private:
    SSolveHandle hnd_;
    std::optional<Model> mdl_;
};

} // namespace

auto Model::symbols(bool shown, bool atoms, bool terms, bool theory) -> SymbolVec {
    auto res = SymbolVec{};
    clingo_show_type_bitset_t show = 0;
    if (shown) {
        show |= clingo_show_type_shown;
    }
    if (atoms) {
        show |= clingo_show_type_atoms;
    }
    if (terms) {
        show |= clingo_show_type_terms;
    }
    if (theory) {
        show |= clingo_show_type_theory;
    }
    handle_error(clingo_model_symbols(mdl_, show, symbol_callback, &res));
    return res;
}

auto Model::str() -> std::string {
    auto res = std::string{};
    auto comma = false;
    for (auto const &sym : symbols(true, false, false, false)) {
        res += comma ? ", " : "";
        res += sym.str();
        comma = true;
    }
    return res;
};

auto SolveHandle::get() -> SolveResult {
    clingo_solve_result_bitset_t res = 0;
    handle_error(clingo_solve_handle_get(hnd_, &res), ptr_);
    return {res};
}

void SolveHandle::cancel() { handle_error(clingo_solve_handle_cancel(hnd_)); }

void SolveHandle::resume() { handle_error(clingo_solve_handle_resume(hnd_)); }

auto SolveHandle::model() -> std::optional<Model> {
    auto const *mdl = static_cast<clingo_model_t const *>(nullptr);
    handle_error(clingo_solve_handle_model(hnd_, &mdl));
    return mdl != nullptr ? std::make_optional<Model>(mdl) : std::nullopt;
}

auto SolveHandle::last() -> std::optional<Model> {
    auto const *mdl = static_cast<clingo_model_t const *>(nullptr);
    handle_error(clingo_solve_handle_last(hnd_, &mdl));
    return mdl != nullptr ? std::make_optional<Model>(mdl) : std::nullopt;
}

auto SolveHandle::core() -> std::vector<clingo_literal_t> {
    auto const *lits = static_cast<clingo_literal_t *>(nullptr);
    auto size = size_t{0};
    handle_error(clingo_solve_handle_core(hnd_, &lits, &size));
    // NOLINTNEXTLINE
    return {lits, lits + size};
}

auto SolveHandle::wait(std::optional<double> timeout) -> bool {
    bool result = false;
    clingo_solve_handle_wait(hnd_, timeout ? *timeout : -1, &result);
    return result;
}

void SolveHandle::close() {
    if (hnd_ != nullptr) {
        clingo_solve_handle_close(std::exchange(hnd_, nullptr));
    }
}

void register_solving(pybind11::module &m) {
    auto solving = m.def_submodule("solving", R"(
Functions and classes related to solving.

Examples
--------

The following example shows how to intercept models with a callback:

    >>> from clingo.core import Library
    >>> from clingo.control import Control
    >>>
    >>> lib = Library()
    >>> ctl = Control(lib, ["0"])
    >>> ctl.parse_string("1 { a; b } 1.")
    >>> ctl.ground()
    >>> with ctl.solve(on_model=print) as hnd:
    ...     print(hnd.get())
    ...
    a
    b
    SAT

The following example shows how to yield models:

    >>> from clingo.core import Library
    >>> from clingo.control import Control
    >>>
    >>> lib = Library()
    >>> ctl = Control(lib, ["0"])
    >>> ctl.parse_string("1 { a; b } 1.")
    >>> ctl.ground()
    >>> with ctl.solve(yield_=True) as hnd:
    ...     for m in hnd:
    ...         print(m)
    ...     print(hnd.get())
    ...
    a
    b
    SAT

The following example shows how to solve asynchronously:

    >>> from clingo.core import Library
    >>> from clingo.control import Control
    >>>
    >>> lib = Library()
    >>> ctl = Control(lib, ["0"])
    >>> ctl.parse_string("1 { a; b } 1.")
    >>> ctl.ground()
    >>> with ctl.solve(on_model=print, async_=True) as hnd:
    ...     # some computation here
    ...     print(hnd.get())
    ...
    a
    b
    SAT

This example shows how to solve both iteratively and asynchronously:

    >>> from clingo.core import Library
    >>> from clingo.control import Control
    >>>
    >>> lib = Library()
    >>> ctl = Control(lib, ["0"])
    >>> ctl.parse_string("1 { a; b } 1.")
    >>> ctl.ground()
    >>> with ctl.solve(yield_=True, async_=True) as hnd:
    ...     while True:
    ...         hnd.resume()
    ...         # some computation here
    ...         m = hnd.model()
    ...         if m is None:
    ...             print(hnd.get())
    ...             break
    ...         print(m)
    ...
    b
    a
    SAT
)"_d);
    py::class_<SolveResult>(solving, "SolveResult", R"(A solve result captures information about a solve call.)")
        .def("__str__", &SolveResult::str, R"(Get a string representation of the solve result.)")
        .def_property_readonly("satisfiable", &SolveResult::satisfiable, R"(Whether there was at least one model.)")
        .def_property_readonly("unsatisfiable", &SolveResult::unsatisfiable, R"(Whether there was no model.)")
        .def_property_readonly("unknown", &SolveResult::unknown, R"(Whether the satisfiablity could be determined.)")
        .def_property_readonly("exhausted", &SolveResult::exhausted, R"(Whether all models have been enumerated.)")
        .def_property_readonly("interrupted", &SolveResult::interrupted, R"(Whether the search was interrupted.)");

    py::class_<Model>(solving, "Model", R"(A view on the solver's current solution.)")
        .def("symbols", &Model::symbols, py::arg("shown") = false, py::arg("atoms") = false, py::arg("terms") = false,
             py::arg("theory") = false, R"(
Get the symbols in the model.

Args:
  shown: Include shown atoms and terms.
  atoms: Include all true atoms including hidden ones.
  terms: Include shown terms.
  theory: Include terms added by external theories.
)"_d)
        .def("__str__", &Model::str, "Get a string representation of the model.");

    py::class_<SolveHandle, SSolveHandle>(solving, "SolveHandle", R"(
An object to interact with a running search.

It can be used to control solving, like, retrieving models or cancelling a
search.

A SolveHandle is a context manager and must be used with Python's with
statement.

Blocking functions in this object release the GIL. They are not thread-safe
though.

See also: `clingo.control.Control.solve`
)"_d)
        .def("get", &SolveHandle::get, R"(
Get the solve result.

This is always the last function that should be called on a handle to ensure
that the search is properly terminated. It might be preceded by a call to
cancel to stop a running search.
)"_d)
        .def("core", &SolveHandle::core, R"(Get the subset of assumptions that made the problem unsatisfiable.)")
        .def("model", &SolveHandle::model, R"(Get the current model if there is any.)")
        .def("last", &SolveHandle::last, R"(
Get the last computed model if there is any.

If the search is not completed yet or the problem is unsatisfiable, the
function returns `None`.
)"_d)
        .def("resume", &SolveHandle::resume, R"(
Discards the last model and starts searching for the next one.

If the search has been started asynchronously, this function starts the search
in the background.
)"_d)
        .def("wait", &SolveHandle::wait, py::arg("timeout") = std::nullopt, R"(
Wait for solve call to finish or the next result with an optional timeout.

If a timeout is given, the behavior of the function changes depending on the
sign of the timeout. If a postive timeout is given, the function blocks for the
given amount time or until a result is ready. If the timeout is negative, the
function will block until a result is ready, which also corresponds to the
behavior of the function if no timeout is given. A timeout of zero can be used
to poll if a result is ready.

Args:
  timeout:
    If a timeout is given, the function blocks for at most timeout seconds.

Returns:
  Indicates whether the solve call has finished or the next result is ready.
)"_d)
        .def("cancel", &SolveHandle::cancel, R"(
Cancel the running search.

See also: `clingo.control.Control.interrupt`
)"_d)
        .def(
            "__enter__", [&](SSolveHandle hnd) -> SSolveHandle { return hnd; }, "Start the search.")
        .def(
            "__exit__",
            [&](SSolveHandle const &hnd, [[maybe_unused]] const std::optional<pybind11::type> &type,
                [[maybe_unused]] const std::optional<pybind11::object> &value,
                [[maybe_unused]] const std::optional<pybind11::object> &traceback) { hnd->close(); },
            "Stop the search closing the handle.")
        .def(
            "__iter__",
            [&](SSolveHandle const &hnd) { return pybind11::make_iterator(ModelIterator{hnd}, ModelIterator{}); },
            "Get an iterator over the models.");
}

} // namespace Clingo::Python
